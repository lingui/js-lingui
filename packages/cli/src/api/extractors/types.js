// @flow

export type ExtractorType = {
  match(filename: string): boolean,
  extract(filename: string, targetDir: string, babelOpts?: Object): void
}
