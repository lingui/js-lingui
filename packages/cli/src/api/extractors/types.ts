export type ExtractorType = {
  match(filename: string): boolean
  extract(filename: string, targetDir: string, options?: ExtractOptions): void
}

export interface BabelOptions {
  plugins?: Array<string>
  presets?: Array<string>
}

export interface ExtractOptions {
  projectType?: string
  babelOptions?: BabelOptions
}
