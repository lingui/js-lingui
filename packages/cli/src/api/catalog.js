// @flow
import fs from "fs"
import path from "path"
import mkdirp from "mkdirp"
import glob from "glob"
import R from "ramda"

import type { LinguiConfig, CatalogApi } from "./types"
import * as locales from "./locales"
import { formats } from "."

export default (config: LinguiConfig): CatalogApi => {
  const format = formats[config.format]
  if (!format) {
    throw new Error(`Unknown format ${config.format}.`)
  }

  const sourceFilename = path.join("{locale}", format.filename)
  const compiledFilename = path.join("{locale}", "messages.js")

  return {
    formatFilename(pattern, locale) {
      return pattern.replace("{locale}", locale)
    },

    write(locale, messages) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      const created = !fs.existsSync(filename)
      format.write(filename, messages, { language: locale })
      return [created, filename]
    },

    read(locale) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) return null
      return format.read(filename)
    },

    readAll() {
      return R.mergeAll(
        this.getLocales().map(locale => ({
          [locale]: this.read(locale)
        }))
      )
    },

    merge(prevCatalogs, nextCatalog, options = {}) {
      const nextKeys = R.keys(nextCatalog)

      return R.mapObjIndexed((prevCatalog, locale) => {
        const prevKeys = R.keys(prevCatalog)

        const newKeys = R.difference(nextKeys, prevKeys)
        const mergeKeys = R.intersection(nextKeys, prevKeys)
        const obsoleteKeys = R.difference(prevKeys, nextKeys)

        // Initialize new catalog with new keys
        const newMessages = R.mapObjIndexed(
          (message, key) => ({
            translation:
              config.sourceLocale === locale ? message.defaults || key : "",
            ...message
          }),
          R.pick(newKeys, nextCatalog)
        )

        // Merge translations from previous catalog
        const mergedMessages = mergeKeys.map(key => {
          const updateFromDefaults =
            config.sourceLocale === locale &&
            (prevCatalog[key].translation === prevCatalog[key].defaults ||
              options.overwrite)

          const translation = updateFromDefaults
            ? nextCatalog[key].defaults
            : prevCatalog[key].translation

          return {
            [key]: {
              translation,
              ...R.omit(["obsolete, translation"], nextCatalog[key])
            }
          }
        })

        // Mark all remaining translations as obsolete
        const obsoleteMessages = obsoleteKeys.map(key => ({
          [key]: {
            ...prevCatalog[key],
            obsolete: true
          }
        }))

        return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
      }, prevCatalogs)
    },

    getTranslation(catalogs, locale, key, { fallbackLocale, sourceLocale }) {
      const getTranslation = locale => catalogs[locale][key].translation

      return (
        // Get translation in target locale
        getTranslation(locale) ||
        // Get translation in fallbackLocale (if any)
        (fallbackLocale && getTranslation(fallbackLocale)) ||
        // Get message default
        catalogs[locale][key].defaults ||
        // If sourceLocale is either target locale of fallback one, use key
        (sourceLocale && sourceLocale === locale && key) ||
        (sourceLocale &&
          fallbackLocale &&
          sourceLocale === fallbackLocale &&
          key) ||
        // Otherwise no translation is available
        undefined
      )
    },

    writeCompiled(locale, content) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(compiledFilename, locale)
      )

      fs.writeFileSync(filename, content)
      return filename
    },

    getLocale(filename) {
      const [messages, locale] = filename.split(path.sep).reverse()

      if (messages !== format.filename || !locales.isValid(locale)) return null
      return locale
    },

    getLocales() {
      const pattern = path.join(config.localeDir, "*")

      return glob
        .sync(pattern)
        .map(filename => {
          // Don't use path.sep here, because glob.sync normalizes path separators
          const [locale] = filename.split("/").reverse()
          return (
            (locales.isValid(locale) || locale === config.pseudoLocale) &&
            locale
          )
        })
        .filter(Boolean)
    },

    addLocale(locale) {
      if (!locales.isValid(locale) && locale !== config.pseudoLocale) {
        return [false, null]
      }

      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) {
        const dirname = path.dirname(filename)

        mkdirp.sync(dirname)
        this.write(locale, {})

        return [true, filename]
      }

      return [false, filename]
    }
  }
}
