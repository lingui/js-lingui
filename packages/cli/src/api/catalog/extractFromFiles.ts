import type {
  ExtractedMessage,
  ExtractorType,
  LinguiConfigNormalized,
} from "@lingui/conf"
import extract from "../extractors"
import path from "path"
import chalk from "chalk"
import { prettyOrigin } from "../utils"
import { MessageOrigin, ExtractedCatalogType } from "../types"

export async function extractFromFiles(
  paths: string[],
  config: LinguiConfigNormalized
) {
  const messages: ExtractedCatalogType = {}

  let catalogSuccess = true
  for (let filename of paths) {
    const fileSuccess = await extract(
      filename,
      (next: ExtractedMessage) => {
        if (!messages[next.id]) {
          messages[next.id] = {
            message: next.message,
            context: next.context,
            comments: [],
            origin: [],
          }
        }

        const prev = messages[next.id]

        // there might be a case when filename was not mapped from sourcemaps
        const filename = next.origin[0]
          ? path.relative(config.rootDir, next.origin[0]).replace(/\\/g, "/")
          : ""

        const origin: MessageOrigin = [filename, next.origin[1]]

        if (prev.message && next.message && prev.message !== next.message) {
          throw new Error(
            `Encountered different default translations for message ${chalk.yellow(
              next.id
            )}` +
              `\n${chalk.yellow(prettyOrigin(prev.origin))} ${prev.message}` +
              `\n${chalk.yellow(prettyOrigin([origin]))} ${next.message}`
          )
        }

        messages[next.id] = {
          ...prev,
          comments: next.comment
            ? [...prev.comments, next.comment]
            : prev.comments,
          origin: [...prev.origin, [filename, next.origin[1]]],
        }
      },
      config,
      {
        extractors: config.extractors as ExtractorType[],
      }
    )
    catalogSuccess &&= fileSuccess
  }

  if (!catalogSuccess) return undefined

  return messages
}
