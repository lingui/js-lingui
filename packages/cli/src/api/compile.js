// @flow
import * as t from "babel-types"
import { parse } from "messageformat-parser"
import { parseExpression } from "babylon"
import generate from "babel-generator"
import plurals from "make-plural"
import R from "ramda"

import type { CatalogType } from "./formats/types"

const isString = s => typeof s === "string"

export function compile(message: string) {
  function processTokens(tokens) {
    if (!tokens.filter(token => !isString(token)).length) {
      return tokens.join("")
    }

    return t.arrayExpression(
      tokens.map(token => {
        if (isString(token)) {
          return t.stringLiteral(token)

          // # in plural case
        } else if (token.type === "octothorpe") {
          return t.stringLiteral("#")

          // simple argument
        } else if (token.type === "argument") {
          return t.callExpression(arg, [t.stringLiteral(token.arg)])

          // argument with custom format (date, number)
        } else if (token.type === "function") {
          const params = [
            t.stringLiteral(token.arg),
            t.stringLiteral(token.key)
          ]

          const format = token.params[0]
          if (format) {
            params.push(t.stringLiteral(format))
          }
          return t.callExpression(arg, params)
        }

        // complex argument with cases
        const formatProps = []

        if (token.offset) {
          formatProps.push(
            t.objectProperty(
              t.identifier("offset"),
              t.numericLiteral(parseInt(token.offset))
            )
          )
        }

        token.cases.forEach(item => {
          const inlineTokens = processTokens(item.tokens)
          formatProps.push(
            t.objectProperty(
              t.identifier(item.key),
              isString(inlineTokens)
                ? t.stringLiteral(inlineTokens)
                : inlineTokens
            )
          )
        })

        const params = [
          t.stringLiteral(token.arg),
          t.stringLiteral(token.type),
          t.objectExpression(formatProps)
        ]

        return t.callExpression(arg, params)
      })
    )
  }

  const arg = t.identifier("a")

  const tokens = parse(message)
  const ast = processTokens(tokens)

  if (isString(ast)) return t.stringLiteral(ast)

  return t.functionExpression(
    null,
    [arg],
    t.blockStatement([t.returnStatement(ast)])
  )
}

export function createCompiledCatalog(locale: string, messages: CatalogType) {
  const [language] = locale.split(/[_-]/)
  const pluralRules = plurals[language]

  const compiledMessages = R.keys(messages).map(key => {
    const translation = messages[key]
    return t.objectProperty(t.stringLiteral(key), compile(translation || key))
  })

  const languageData = [
    t.objectProperty(
      t.stringLiteral("plurals"),
      parseExpression(pluralRules.toString())
    )
  ]

  const ast = t.expressionStatement(
    t.assignmentExpression(
      "=",
      t.memberExpression(t.identifier("module"), t.identifier("exports")),
      t.objectExpression([
        // language data
        t.objectProperty(
          t.identifier("languageData"),
          t.objectExpression(languageData)
        ),
        // messages
        t.objectProperty(
          t.identifier("messages"),
          t.objectExpression(compiledMessages)
        )
      ])
    )
  )

  return generate(ast, {
    minified: true
  }).code
}
