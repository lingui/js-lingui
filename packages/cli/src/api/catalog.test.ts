import fs from "fs"
import path from "path"
import mockFs from "mock-fs"
import { mockConsole } from "@lingui/jest-mocks"
import { LinguiConfig, makeConfig } from "@lingui/conf"

import { Catalog, cleanObsolete, order } from "./catalog"
import { createCompiledCatalog } from "./compile"

import {
  copyFixture,
  defaultMakeOptions,
  defaultMakeTemplateOptions,
  makeNextMessage,
  defaultMergeOptions,
  makeCatalog,
} from "../tests"
import { AllCatalogsType } from "./types"
import { extractFromFiles } from "./catalog/extractFromFiles"
import { FormatterWrapper, getFormat } from "./formats"

export const fixture = (...dirs: string[]) =>
  path.resolve(__dirname, path.join("fixtures", ...dirs)) +
  // preserve trailing slash
  (dirs[dirs.length - 1].endsWith("/") ? "/" : "")

function mockConfig(config: Partial<LinguiConfig> = {}) {
  return makeConfig(
    {
      rootDir: path.join(__dirname, "fixtures"),
      locales: ["en", "pl"],
      ...config,
    },
    { skipValidation: true }
  )
}

describe("Catalog", () => {
  let format: FormatterWrapper

  beforeAll(async () => {
    format = await getFormat("po", {}, "en")
  })

  afterEach(() => {
    mockFs.restore()
  })

  describe("make", () => {
    it("should collect and write catalogs", async () => {
      const localeDir = await copyFixture(fixture("locales", "initial"))
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.join(localeDir, "{locale}", "messages"),
          include: [
            fixture("collect/componentA/"),
            fixture("collect/componentB"),
          ],
          exclude: [],
          format,
        },
        mockConfig({
          locales: ["en", "cs"],
        })
      )

      // Everything should be empty
      expect(await catalog.readAll()).toMatchSnapshot()

      await catalog.make(defaultMakeOptions)
      expect(await catalog.readAll()).toMatchSnapshot()
    })

    it("should only update the specified locale", async () => {
      const localeDir = await copyFixture(fixture("locales", "initial"))
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.join(localeDir, "{locale}", "messages"),
          include: [
            fixture("collect/componentA/"),
            fixture("collect/componentB"),
          ],
          exclude: [],
          format,
        },
        mockConfig({
          locales: ["en", "cs"],
        })
      )

      // Everything should be empty
      expect(await catalog.readAll()).toMatchSnapshot()

      await catalog.make({ ...defaultMakeOptions, locale: "en" })
      expect(await catalog.readAll()).toMatchSnapshot()
    })

    it("should merge with existing catalogs", async () => {
      const localeDir = await copyFixture(fixture("locales", "existing"))
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.join(localeDir, "{locale}"),
          include: [fixture("collect/")],
          exclude: [],
          format,
        },
        mockConfig({
          locales: ["en", "cs"],
        })
      )

      // Everything should be empty
      expect(await catalog.readAll()).toMatchSnapshot()

      await catalog.make(defaultMakeOptions)
      expect(await catalog.readAll()).toMatchSnapshot()
    })
  })

  describe("makeTemplate", () => {
    it("should collect and write a template", async () => {
      const localeDir = await copyFixture(fixture("locales", "initial"))
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.join(localeDir, "{locale}", "messages"),
          include: [
            fixture("collect/componentA/"),
            fixture("collect/componentB"),
          ],
          exclude: [],
          format,
        },
        mockConfig({
          locales: ["en", "cs"],
        })
      )

      // Everything should be empty
      expect(await catalog.readTemplate()).toMatchSnapshot()

      await catalog.makeTemplate(defaultMakeTemplateOptions)
      expect(await catalog.readTemplate()).toMatchSnapshot()
    })
  })

  describe("POT Flow", () => {
    it("Should get translations from template if locale file not presented", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.resolve(
            __dirname,
            path.join("fixtures", "pot-template", "{locale}")
          ),
          include: [],
          exclude: [],
          format,
        },
        mockConfig({
          locales: ["en", "pl"],
        })
      )

      const translations = await catalog.getTranslations("en", {
        sourceLocale: "en",
        fallbackLocales: {
          default: "en",
        },
      })

      expect(translations).toMatchSnapshot()
    })
  })

  describe("collect", () => {
    it("should support JSX and Typescript", async () => {
      const messages = await extractFromFiles(
        [
          fixture("collect-typescript-jsx/jsx-in-js.js"),
          fixture("collect-typescript-jsx/jsx-syntax.jsx"),
          fixture("collect-typescript-jsx/tsx-syntax.tsx"),
          fixture("collect-typescript-jsx/macro.tsx"),
        ],
        mockConfig()
      )

      expect(messages).toMatchSnapshot()
    })

    it("should support experimental typescript decorators under a flag", async () => {
      const messages = await extractFromFiles(
        [fixture("collect-typescript-jsx/tsx-experimental-decorators.tsx")],
        mockConfig({
          extractorParserOptions: {
            tsExperimentalDecorators: true,
          },
        })
      )

      expect(messages).toBeTruthy()
      expect(messages).toMatchInlineSnapshot(`
        {
          xDAtGP: {
            comments: [],
            context: undefined,
            message: Message,
            origin: [
              [
                collect-typescript-jsx/tsx-experimental-decorators.tsx,
                15,
              ],
            ],
          },
        }
      `)
    })

    it("should support Flow syntax if enabled", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [fixture("collect-syntax-flow/")],
          exclude: [],
          format,
        },
        mockConfig({
          extractorParserOptions: {
            flow: true,
          },
        })
      )

      const messages = await catalog.collect()
      expect(messages).toMatchSnapshot()
    })
    it("should extract messages from source files", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [fixture("collect/")],
          exclude: [],
          format,
        },
        mockConfig()
      )

      const messages = await catalog.collect()
      expect(messages).toMatchSnapshot()
    })

    it("should respect inline sourcemaps", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [fixture("collect-inline-sourcemaps/")],
          exclude: [],
          format,
        },
        mockConfig()
      )

      const messages = await catalog.collect()
      expect(messages[Object.keys(messages)[0]].origin).toStrictEqual([
        ["../../../../../input.tsx", 5],
      ])
    })

    it("should extract only files passed on options", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [
            fixture("collect/componentA"),
            fixture("collect/componentB.js"),
          ],
          exclude: [],
          format,
        },
        mockConfig()
      )

      const messages = await catalog.collect({
        files: [fixture("collect/componentA")],
      })
      expect(messages).toMatchSnapshot()
    })

    it("should throw an error when duplicate identifier with different defaults found", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [fixture("duplicate-id.js")],
          exclude: [],
          format,
        },
        mockConfig()
      )
      expect.assertions(1)
      await mockConsole(async (console) => {
        await catalog.collect({
          files: [fixture("duplicate-id.js")],
        })

        expect(console.error).toBeCalledWith(
          expect.stringContaining(
            `Encountered different default translations for message`
          )
        )
      })
    })

    it("should handle syntax errors", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [fixture("collect-invalid/")],
          exclude: [],
          format,
        },
        mockConfig()
      )

      expect.assertions(2)
      await mockConsole(async (console) => {
        const messages = await catalog.collect()
        expect(console.error).toBeCalledWith(
          expect.stringContaining(`Cannot process file`)
        )
        expect(messages).toBeFalsy()
      })
    })
  })
  it("Catalog.merge should initialize catalogs", async () => {
    const prevCatalogs: AllCatalogsType = { en: null, cs: null }
    const nextCatalog = {
      "custom.id": makeNextMessage({
        message: "Message with custom ID",
      }),
      "Message with <0>auto-generated</0> ID": makeNextMessage(),
    }

    expect(
      (await makeCatalog({ sourceLocale: "en" })).merge(
        prevCatalogs,
        nextCatalog,
        defaultMergeOptions
      )
    ).toEqual({
      // catalog for sourceLocale - translation is prefilled
      en: {
        "custom.id": expect.objectContaining({
          message: "Message with custom ID",
          translation: "Message with custom ID",
        }),
        "Message with <0>auto-generated</0> ID": expect.objectContaining({
          translation: "Message with <0>auto-generated</0> ID",
        }),
      },
      // catalog for other than sourceLocale - translation is empty
      cs: {
        "custom.id": expect.objectContaining({
          message: "Message with custom ID",
          translation: "",
        }),
        "Message with <0>auto-generated</0> ID": expect.objectContaining({
          translation: "",
        }),
      },
    })
  })

  describe("read", () => {
    it("should return null if file does not exist", async () => {
      // mock empty filesystem
      mockFs()

      const catalog = new Catalog(
        {
          name: "messages",
          path: "locales/{locale}",
          include: [],
          exclude: [],
          format,
        },
        mockConfig()
      )

      const messages = await catalog.read("en")
      expect(messages).toBeNull()
    })

    it("should read file in given format", async () => {
      mockFs({
        en: {
          "messages.po": fs.readFileSync(
            path.resolve(__dirname, "fixtures/messages.po")
          ),
        },
      })
      const catalog = new Catalog(
        {
          name: "messages",
          path: "{locale}/messages",
          include: [],
          format,
        },
        mockConfig()
      )

      const messages = await catalog.read("en")

      mockFs.restore()
      expect(messages).toMatchSnapshot()
    })

    xit("should read file in previous format", async () => {
      mockFs({
        en: {
          "messages.json": fs.readFileSync(
            path.resolve(__dirname, "fixtures/messages.json")
          ),
        },
      })
      const catalog = new Catalog(
        {
          name: "messages",
          path: "{locale}/messages",
          include: [],
          format,
        },
        mockConfig({ prevFormat: "minimal" })
      )

      const messages = await catalog.read("en")

      mockFs.restore()
      expect(messages).toMatchSnapshot()
    })
  })

  describe("readAll", () => {
    it("should read existing catalogs for all locales", async () => {
      const catalog = new Catalog(
        {
          name: "messages",
          path: path.resolve(
            __dirname,
            path.join("fixtures", "readAll", "{locale}", "messages")
          ),
          include: [],
          format,
        },
        mockConfig({
          locales: ["en", "cs"],
        })
      )

      const messages = await catalog.readAll()
      expect(messages).toMatchSnapshot()
    })
  })
})

describe("cleanObsolete", () => {
  it("should remove obsolete messages from catalog", () => {
    const catalog = {
      Label: makeNextMessage({
        translation: "Label",
      }),
      PreviousLabel: makeNextMessage({
        obsolete: true,
      }),
    }

    expect(cleanObsolete(catalog)).toMatchSnapshot()
  })
})

describe("order", () => {
  it("should order messages alphabetically", () => {
    const catalog = {
      LabelB: makeNextMessage({
        translation: "B",
      }),
      LabelA: makeNextMessage({
        translation: "A",
      }),
      LabelD: makeNextMessage({
        translation: "D",
      }),
      LabelC: makeNextMessage({
        translation: "C",
      }),
    }

    const orderedCatalogs = order("messageId")(catalog)

    // Test that the message content is the same as before
    expect(orderedCatalogs).toMatchSnapshot()

    // Jest snapshot order the keys automatically, so test that the key order explicitly
    expect(Object.keys(orderedCatalogs)).toMatchSnapshot()
  })

  it("should order messages by origin", () => {
    const catalog = {
      LabelB: makeNextMessage({
        translation: "B",
        origin: [
          ["file2.js", 2],
          ["file1.js", 2],
        ],
      }),
      LabelA: makeNextMessage({
        translation: "A",
        origin: [["file2.js", 3]],
      }),
      LabelD: makeNextMessage({
        translation: "D",
        origin: [["file2.js", 100]],
      }),
      LabelC: makeNextMessage({
        translation: "C",
        origin: [["file1.js", 1]],
      }),
    }

    const orderedCatalogs = order("origin")(catalog)

    // Test that the message content is the same as before
    expect(orderedCatalogs).toMatchSnapshot()

    // Jest snapshot order the keys automatically, so test that the key order explicitly
    expect(Object.keys(orderedCatalogs)).toMatchSnapshot()
  })

  it("should order messages by message", () => {
    const catalog = {
      msg1: makeNextMessage({
        message: "B",
        translation: "B",
        origin: [
          ["file2.js", 2],
          ["file1.js", 2],
        ],
      }),
      msg2: makeNextMessage({
        // message is optional.
        translation: "A",
        origin: [["file2.js", 3]],
      }),
      msg3: makeNextMessage({
        message: "D",
        translation: "D",
        origin: [["file2.js", 100]],
      }),
      msg4: makeNextMessage({
        message: "C",
        translation: "C",
        origin: [["file1.js", 1]],
      }),
    }

    const orderedCatalogs = order("message")(catalog)

    // Jest snapshot order the keys automatically, so test that the key order explicitly
    expect(Object.keys(orderedCatalogs)).toMatchInlineSnapshot(`
      [
        msg2,
        msg1,
        msg4,
        msg3,
      ]
    `)
  })
})

describe("writeCompiled", () => {
  let catalog: Catalog

  beforeAll(async () => {
    const localeDir = await copyFixture(fixture("locales", "initial/"))
    catalog = new Catalog(
      {
        name: "messages",
        path: path.join(localeDir, "{locale}", "messages"),
        include: [],
        exclude: [],
        format: await getFormat("po", {}, "en"),
      },
      mockConfig()
    )
  })

  it.each([
    { namespace: "es", extension: /\.mjs$/ },
    { namespace: "ts", extension: /\.ts$/ },
    { namespace: undefined, extension: /\.js$/ },
    { namespace: "cjs", extension: /\.js$/ },
    { namespace: "window.test", extension: /\.js$/ },
    { namespace: "global.test", extension: /\.js$/ },
  ])(
    "Should save namespace $namespace in $extension extension",
    async ({ namespace, extension }) => {
      const compiledCatalog = createCompiledCatalog("en", {}, { namespace })
      // Test that the file extension of the compiled catalog is `.mjs`
      expect(
        await catalog.writeCompiled("en", compiledCatalog, namespace)
      ).toMatch(extension)
    }
  )
})
