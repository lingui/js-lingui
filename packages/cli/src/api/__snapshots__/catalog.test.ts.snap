// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog collect should extract messages from source files 1`] = `
Object {
  Hello World: Object {
    origin: Array [
      Array [
        collect/componentA/componentA.js,
        1,
      ],
      Array [
        collect/componentA/index.js,
        1,
      ],
    ],
  },
}
`;

exports[`Catalog collect should handle errors 1`] = `Object {}`;

exports[`Catalog make should collect and write catalogs 1`] = `
Object {
  cs: null,
  en: null,
}
`;

exports[`Catalog make should collect and write catalogs 2`] = `
Object {
  cs: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [
        Array [
          collect/componentA/componentA.js,
          1,
        ],
        Array [
          collect/componentA/index.js,
          1,
        ],
      ],
      translation: ,
    },
  },
  en: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [
        Array [
          collect/componentA/componentA.js,
          1,
        ],
        Array [
          collect/componentA/index.js,
          1,
        ],
      ],
      translation: ,
    },
  },
}
`;

exports[`Catalog make should merge with existing catalogs 1`] = `
Object {
  cs: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [],
      translation: Ahoj Brno,
    },
  },
  en: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [],
      translation: Ahoj Brno,
    },
  },
}
`;

exports[`Catalog make should merge with existing catalogs 2`] = `
Object {
  cs: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [
        Array [
          collect/componentA/componentA.js,
          1,
        ],
        Array [
          collect/componentA/index.js,
          1,
        ],
      ],
      translation: Ahoj Brno,
    },
  },
  en: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [
        Array [
          collect/componentA/componentA.js,
          1,
        ],
        Array [
          collect/componentA/index.js,
          1,
        ],
      ],
      translation: Ahoj Brno,
    },
  },
}
`;

exports[`Catalog read should read file in given format 1`] = `
Object {
  obsolete: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: true,
    origin: Array [],
    translation: Is marked as obsolete,
  },
  static: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Static message,
  },
  veryLongString: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. "What's happened to me?" he thought. It wasn't a dream. His room, a proper human,
  },
  withComments: Object {
    comment: undefined,
    comments: Array [
      Translator comment,
      This one might come from developer,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Support translator comments separately,
  },
  withDescription: Object {
    comment: Description is comment from developers to translators,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Message with description,
  },
  withFlags: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [
      fuzzy,
      otherFlag,
    ],
    obsolete: false,
    origin: Array [],
    translation: Keeps any flags that are defined,
  },
  withMultipleOrigins: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
      Array [
        src/Component.js,
        2,
      ],
    ],
    translation: Message with multiple origin,
  },
  withOrigin: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
    ],
    translation: Message with origin,
  },
}
`;

exports[`Catalog read should read file in previous format 1`] = `null`;

exports[`Catalog readAll should read existing catalogs for all locales 1`] = `
Object {
  cs: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [],
      translation: Ahoj Brno,
    },
  },
  en: Object {
    Hello World: Object {
      comment: undefined,
      comments: Array [],
      flags: Array [],
      obsolete: false,
      origin: Array [],
      translation: Hello World,
    },
  },
}
`;

exports[`Catalog should convert catalog format 1`] = `
msgid ""
msgstr ""
"Project-Id-Version: \\n"
"Report-Msgid-Bugs-To: \\n"
"POT-Creation-Date: \\n"
"PO-Revision-Date: \\n"
"Last-Translator: \\n"
"Language: \\n"
"Language-Team: \\n"
"Content-Type: \\n"
"Content-Transfer-Encoding: \\n"
"Plural-Forms: \\n"

msgid "static"
msgstr "Static message"

msgid "veryLongString"
msgstr "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. \\"What's happened to me?\\" he thought. It wasn't a dream. His room, a proper human"
`;

exports[`cleanObsolete should remove obsolete messages from catalog 1`] = `
Object {
  Label: Object {
    obsolete: false,
    origin: Array [
      Array [
        catalog.test.ts,
        1,
      ],
    ],
    translation: Label,
  },
}
`;

exports[`getCatalogs should warn about missing {name} pattern in catalog path 1`] = `Catalog with path "./locales/{locale}" doesn't have a {name} pattern in it, but one of source directories uses it: "{name}/". Either add {name} pattern to "./locales/{locale}" or remove it from all source directories.`;

exports[`getCatalogs should warn if catalogPath is a directory 1`] = `Remove trailing slash from "./locales/{locale}/". Catalog path isn't a directory, but translation file without extension. For example, catalog path "./locales/{locale}" results in translation file "./locales/en.po".`;

exports[`getCatalogs should warn if catalogPath is a directory 2`] = `Remove trailing slash from "./locales/{locale}/". Catalog path isn't a directory, but translation file without extension. For example, catalog path "./locales/{locale}" results in translation file "./locales/cs.json".`;

exports[`order should order messages alphabetically 1`] = `
Object {
  LabelA: Object {
    obsolete: false,
    origin: Array [
      Array [
        catalog.test.ts,
        1,
      ],
    ],
    translation: A,
  },
  LabelB: Object {
    obsolete: false,
    origin: Array [
      Array [
        catalog.test.ts,
        1,
      ],
    ],
    translation: B,
  },
  LabelC: Object {
    obsolete: false,
    origin: Array [
      Array [
        catalog.test.ts,
        1,
      ],
    ],
    translation: C,
  },
  LabelD: Object {
    obsolete: false,
    origin: Array [
      Array [
        catalog.test.ts,
        1,
      ],
    ],
    translation: D,
  },
}
`;

exports[`order should order messages alphabetically 2`] = `
Array [
  LabelA,
  LabelB,
  LabelC,
  LabelD,
]
`;

exports[`order should order messages by origin 1`] = `
Object {
  LabelA: Object {
    obsolete: false,
    origin: Array [
      Array [
        file2.js,
        3,
      ],
    ],
    translation: A,
  },
  LabelB: Object {
    obsolete: false,
    origin: Array [
      Array [
        file1.js,
        2,
      ],
      Array [
        file2.js,
        2,
      ],
    ],
    translation: B,
  },
  LabelC: Object {
    obsolete: false,
    origin: Array [
      Array [
        file1.js,
        1,
      ],
    ],
    translation: C,
  },
  LabelD: Object {
    obsolete: false,
    origin: Array [
      Array [
        file2.js,
        100,
      ],
    ],
    translation: D,
  },
}
`;

exports[`order should order messages by origin 2`] = `
Array [
  LabelC,
  LabelB,
  LabelA,
  LabelD,
]
`;
